lte/gateway/c/session_manager/SessionState.cpp

```diff
@@ -235,26 +235,64 @@ uint32_t SessionState::get_current_version() {
 }
 
 void SessionState::set_current_version(
-    int new_session_version, SessionStateUpdateCriteria* session_uc) {
-  current_version_ = new_session_version;
-  if (session_uc) {
-    session_uc->is_current_version_updated = true;
-    session_uc->updated_current_version    = new_session_version;
-  }
+    int new_session_version, SessionStateUpdateCriteria& session_uc) {
+  current_version_                      = new_session_version;
+  session_uc.is_current_version_updated = true;
+  session_uc.updated_current_version    = new_session_version;
   MLOG(MINFO) << " Current version is " << get_current_version();
 }
+
 /* Add PDR rule to this rules session list */
-void SessionState::insert_pdr(SetGroupPDR* rule) {
-  PdrList_.push_back(*rule);
+void SessionState::insert_pdr(
+    SetGroupPDR* rule, SessionStateUpdateCriteria& session_uc, bool crit_add) {
+  // Check if it already exists
+  int32_t Pdr_index;
+  Pdr_index = pdr_index(rule->pdr_id());
+  if (Pdr_index != -1) {
+    // Update the existing value
+    PdrList_.at(Pdr_index) = *rule;
+  } else {
+    // Insert the rule
+    PdrList_.push_back(*rule);
+  }
+  // update criteria to be updated
+  if (crit_add) session_uc.pdrs_to_install.push_back(*rule);
+}
+
+/* method to change the PDR state */
+void SessionState::set_all_pdrs(enum PdrState pdr_state) {
+  for (auto& rule : PdrList_) {
+    rule.set_pdr_state(pdr_state);
+  }
+}
+
+int32_t SessionState::pdr_index(uint32_t id) {
```

Comment:
instead of returning -1 on failure, we can return `optional<uint32_t>`

Reply:
It's getting compilation because index number used in a vector. If possible, I will do fix later.